add `same_line` and `anchored` movements

pantos9000 opened the pull request:
This PR picks up the stale PR #4204. Big thanks to @mitsuhiko for his PR. Seeing the old changes made it really easy to jump right into the code :sunglasses: :+1: 

This PR adds two kinds of new movement sets: `same_line` and `anchored` movement.

For more info, check the old PR #4204.

## `same_line` movement

Commands added:
* `move_same_line_char_left`
* `move_same_line_char_right`
* `extend_same_line_char_left`
* `extend_same_line_char_right`
* `append_mode_same_line`

These new commands move cursors, while making them stay in the same line. So if a cursor would wrap around into another line, instead it won't move and stay at its current position.

## `anchored` movement

Commands added:
* `move_anchored_line_up`
* `move_anchored_line_down`
* `move_anchored_visual_line_up`
* `move_anchored_visual_line_down`
* `extend_anchored_line_up`
* `extend_anchored_line_down`
* `extend_anchored_visual_line_up`
* `extend_anchored_visual_line_down`

These new commands move cursors vertically. A cursor will move depending on its position:

* If it is on a newline character of a non-empty line, the cursor will stay on newlines (i.e. on a line's last character).
* If it is on a non-newline character of a non-empty line, the cursor will try to avoid newline characters. It will move normally, but if it would end up on a newline, instead it will be moved one position left of it (i.e. the line's second to last character).
* If it is on the newline character of an empty line (that contains nothing except the newline character), the cursor will continue to move like before: If it stayed on newline before, it will continue to do so. Otherwise it will try to avoid them (except on empty lines).

## Differences to the old PR
* Based on recent master
* Commands support soft-wrap / visual movement
* Added some unit tests
* Slightly changed command names
* Additional command `append_mode_same_line` (for consistency, as append is more or less move right + insert mode)

## Remarks
#### Command / function names
I was not sure about the general naming - e.g. `anchored` I think is used somewhere else in the code in a different context. I guess it would be bad if this somehow collided, so you might want to keep that in mind during review. If anyone spots any ambiguities - I'm always open for improvements :)

#### Helper functions context
I created some functions that are only used locally, e.g. `is_in_visual_empty_line()`. In case you think it makes sense to have some of those in a more common context (core?), I can gladly move them.

#### Unit tests
I struggled a bit for the soft-wrap functionality, the corresponding tests might be too implementation-specific and break in the future, if the soft-wrap functionality is adjusted. Couldn't find a better way, so I hope the tests are okay like this.

Also I skipped the unit tests for `Movement::Extend`, as the functionality is exactly the same as with `Movement::Move`.

#### Docs
I don't know if these commands should be documented somewhere, e.g. in the book. If it makes sense, I can write a few describing words.

#### Only a single setting
If these movements should have only a single setting (i.e. a bool that can be set) then this would probably affect a bit the architecture. Not sure how it would be best to implement it, but I would maybe go for additional values in either `helix_core::Movement` or `helix_core::Direction`.

But I think the way it was implemented previously (and still is in this PR) allows for quite versatile configuration and fits nicely into the existing architecture, so I would prefer it the way it is.

Edit: Additional remarks, typos

pantos9000 committed:
add `same_line` movement commands (#768)

Commands added:
* `move_same_line_char_left`
* `move_same_line_char_right`
* `extend_same_line_char_left`
* `extend_same_line_char_right`
* `append_mode_same_line`

These new commands move cursors, while making them stay in the same
line. So if a cursor would wrap around into another line, instead it
won't move and stay at its current position.

pantos9000 committed:
add `anchored` movement commands

Commands added:
* `move_anchored_line_up`
* `move_anchored_line_down`
* `move_anchored_visual_line_up`
* `move_anchored_visual_line_down`
* `extend_anchored_line_up`
* `extend_anchored_line_down`
* `extend_anchored_visual_line_up`
* `extend_anchored_visual_line_down`

These new commands move cursors vertically. A cursor will move depending
on its position:

* If it is on a newline character of a non-empty line, the cursor will
  stay on newlines (i.e. on a line's last character).

* If it is on a non-newline character of a non-empty line, the cursor
  will try to avoid newline characters. It will move normally, but if
  it would end up on a newline, instead it will be moved one position
  left of it (i.e. the line's second to last character).

* If it is on the newline character of an empty line (that contains
  nothing except the newline character), the cursor will continue to
  move like before: If it stayed on newline before, it will continue to
  do so. Otherwise it will try to avoid them (except on empty lines).

Pantos9000 commented:
Found a bug in the `anchored` implementation, made the PR a draft again.

mitsuhiko commented:
Thank you so much for picking this back up.

pantos9000 commented:
To avoid misunderstandings: Just wanted to let people know that from my side, this PR is ready for review. Let me know if anything is missing, will gladly put additional work into this if needed.

pantos9000 commented:
I will look into the PR from time to time and try to rebase it on recent master. As this is open since a few months now, I just wanted to ask: Is this PR still wanted?

kirawi commented:
The maintainers have been busy outside of Helix so there is quite a backlog of PRs to review. It'll happen when it'll happen.

hholst80 commented:
I have tried this PR and it an absolutely a home-run for me!

1. Clone the branch https://github.com/pantos9000/helix/tree/feature/new_movements
2. `cargo build --release`
3. `cp build/target/hx /usr/local/bin` (override installed `hx`)
4. start `hx` and run `:open-config`
5. add config

Relevant config:

```toml
[keys]
[keys.normal]
"h" = "move_same_line_char_left"
"l" = "move_same_line_char_right"
```

pantos9000 commented:
> Hopefully this was clear enough, it's difficult to describe this stuff.

Wow, nice bug report. Was very easy to follow, thanks :)

I can reproduce both bugs, will try to fix them the coming days. I did never test append in visual/select mode, so thanks for catching it!

pantos9000 committed:
fix command `append_mode_same_line`

If the cursor was not at the end of the selection range, but the range
end was a newline, `append_mode_same_line` would cause the cursor to
jump to a wrong position.

See https://github.com/helix-editor/helix/pull/10576#issuecomment-2990522009
for details.

pantos9000 commented:
Added a fix to the last fix - when at the end of the document, a `debug_assert!()` would trigger.

Fixup'ed it to keep things tidy.

pantos9000 commented:
I tried to implement unit tests for the new append command, but struggled a lot with it. I failed to find a lean way to create the context that the function acts upon, without having to create the whole editor context. Maybe there is an easier way, but I failed to find it.

If a maintainer reads this and would like to see those tests being added, a hint in the right direction would be appreciated :)

pantos9000 committed:
generate docs for new movement commands
